### PR TITLE
Updated JoinHashTable to use linear probing to resolve hash collisions

### PR DESCRIPTION
**Pull request by gropaul**
# TL;DR

In this PR, the Join HashTable is adapted to use linear probing to resolve hash collisions. The key changes are 
- ensure the chains of tuples for one HashTable entry share a unique key, so we only need to compare the first elements of a chain, which reduces the number of comparisons if there are duplicate keys on the build side.
- using salts in the HashTable Pointers similar to the Aggregate HashTable as a very performant preliminary comparison without cache misses

Benchmarking the new implementation led to the following runtime improvements for the TPCH Benchmark on a MacBook Pro with an M1 chip using 6 threads:

| Scale Factor  | 1 | 3 | 10 | 30 | 100 |
| --- | --- | --- | --- | --- | --- |
| Existing Implementation | 0.50165 | 1.34258 | 4.73738 | 14.5952 | 55.672 |
| Linear Probing (new) | 0.47907 | 1.27873 | 4.43729 | 13.6132 | 51.8013 |
| Relative Performance Increase | 4.50% | 4.75% | 6.33% | 6.72% | 6.95% |

# How it works

### Initialization

To ensure that there are different keys in the Chain of one HashTable Entry, when inserting a key into an already full entry we need to compare it with the key already in the HashTable. Therefore, when creating the HashTable, a new RowMatcher is initialized which is used during building and also during the probing phase.

For the existing implementation, non-equality predicates are also utilized during the `ResolvePredicates` step. However, these cannot be applied in the Matcher used during building. For instance, if there were an "is-larger-than" comparison for the keys, the keys inserted into one entry would vary based on their order of insertion. For example, if the keys 3, 4, 2 were inserted into the same entry, three would be the first element against which the others would be compared, so 2 would not end up in the chain. Conversely, if they were ordered 2, 3, 4, all keys would end up in the chain of the same entry.

This is why there are now two types of RowMatcher, one that is used during build and probing containing all equality predicates and one that is used during the `ResolvePredicates` step containing the remaining conditions. In the initialization phase, the corresponding predicates and types for both RowMatchers are collected.

### Insertion

During insertion, for each key that is to be inserted, the offset of the key into the hash table is calculated to check if the corresponding entry is empty.

If it is found to be empty, the salt of the key's hash, along with a pointer pointing to the row of the respective key, is inserted into the entry. When insertions occur in parallel, it is possible that while key A is being inserted into entry E, another thread might also attempt to insert key B into the same entry E. In the previous implementation, having different keys in the chain of entry was allowed, but this needs to be prevented with the new implementation.  Therefore, a Compare and Swap operation is employed, which only swaps if the entry remains empty. If this is not the case, the key is added to a list of keys that require comparison against their entry.

If the entry targeted for a key insertion is full, the salts are compared. Should the salts differ, the search moves to the next entry. This process is repeated until an empty entry or an entry with matching salts is found. Upon finding a match in salts, the key within the entry is compared. If a match is found, the new key is appended to the entry. Otherwise, the search moves to the next hash table entry, starting the process new.

While comparing the tuples during insertion, there was the problem of having the RowLayout on both the left and right side, while the RowMatcher is build to compare a DataChunk against the RowLayout. Therefore, the data for the keys to be inserted is first gathered into a DataChunk and than compared. This is an overhead that could be resolved in the future. 

### Probing

The probing phase of the HashTable has been modified to do more than just resolve pointers to the hash table entries; it now also checks for matches and directly returns the pointer to the entry's chain. This process is managed by the `GetRowPointers` function, which additionally returns a `SelectionVector` and a count of the matches.

The `GetRowPointers` can be run in two modes: Using salt or not using salt. This was added as for small HashTables, the chance of both HashCollisions for keys and cache misses while comparing the incoming keys with the rows inside the HashTable is lower. E.g. if there are only 5 (distinct) keys in the HashTable, but you probe for a million tuples, the salt comparison will always be correct. The current metric for not using the salt is if the HashTable is smaller or equal than 8192 and if there is only one equality predicate. The latter was added because, for example, if there are 5 predicates that need to be checked for every insertion, comparing the salt is faster. This is because the hash, which the salt is based on, is itself derived from those 5 predicates.

As the overall logic of the Probing is now more complex than in the previous implementation (calculating and comparing salt, comparing keys), one big optimization was to hide this logic from keys where there are no matches in the HashTable. This is why first, the incoming data chunk is filtered for all keys that actually have a match in the HashTable. 

For all keys that result in a hit, and if the salt comparison is enabled, the salt is calculated using their hash. Then, for each key to be probed, the salt of the respective entry is checked. If the salt does not match, the search continues to the next entry until either an empty entry is found or one where the salt does match. In the case of finding an empty entry, there is no match and the key is not processed any further. If a salt match is found, the probe key is compared against the key of the first element in the list of the HashTable entry. Should the comparison be successful, a pointer to the list of the entry is returned, and the probe key is added to the return `SelectionVector`. If there is no match, the next entry is checked and the loop starts again. 

### Other Changes: `ScanStructure::ResolvePredicates`

As discussed in the Initialisation section, there are now two RowMatchers, one during build that uses all the equality predicates, and that keeps all the remaining predicates. During the `ScanStructure::ResolvePredicates` function, if there are remaining predicates, a match is performed and the count of matches is returned. If not, the function exists directly.  

### Other Changes: `RowMatcher::Initialize` and `RowMatcher::Match`

The RowMatcher was altered to support the option to only match against some of the columns. This is needed as we now have two matches that compare against a distinct set of predicates: The equality and the non-equal predicates. 

(I tried slicing the DataChunk to not make this alternation to the Matcher, but initialising the RowMatcher with the subset of column types and also altering the `TupleDataLayout` turned out to be complicated.) 

### Performance Evaluation

To evaluate the performance of the TPCH Benchmark was run on scale factors of 1, 3, 10, 30, and 100 on a Mac Book Pro using 6 threads:

Chip:         Apple M1
Cores:       8 (4 performance and 4 efficiency)
Memory:	 16 GB
OS:            macOS Ventura 13.6

The plot below shows the total runtime for the TPCH benchmarks. For scale factor 100, the benchmark runner had a timeout for queries 7 and 18 for the existing implementation and on query 18 for the new implementation. This is indicated by a negative runtime and a relative runtime of 0.

query_runtimes

If there are any questions or change requests, let me know! :)

### CODE REVIEW

File: src/execution/join_hashtable.cpp [hunk 1]
```diff
@@ -86,32 +120,196 @@ void JoinHashTable::Merge(JoinHashTable &other) {
...
+		remaining_sel = &state.key_no_match_sel;
+		remaining_count = key_no_match_count;
+	}
+}
+
+inline bool JoinHashTable::UseSalt() const {
+	// only compare salts with the ht entries if the capacity is larger than 8192 so
+	// that it does not fit into the CPU cache and if there is only one equality condition as otherwise
+	// we potentially need to compare multiple keys
+	return this->capacity > 8192 && this->equality_predicate_columns.size() == 1;
```
Comment (lnkuiper) [inline src/execution/join_hashtable.cpp:300]: Could the `8192` be a `static constexpr const idx_t USE_SALT_THRESHOLD = 8192;` defined the header with this descriptive comment there instead?

File: src/execution/join_hashtable.cpp [hunk 2]
```diff
@@ -243,70 +441,249 @@ idx_t JoinHashTable::PrepareKeys(DataChunk &keys, vector<TupleDataVectorFormat>
...
+			} else { // compare with full
+				state.salt_match_sel.set_index(salt_match_count, row_index);
+				row_ptr_insert_to[salt_match_count] = entry.GetPointer();
+				salt_match_count += 1;
+			}
+		}
+
+		// at this step, for all the pointers_result_v we stepped either until we found an empty entry or an entry with
+		// a matching salt, we need to compare the keys for the ones that have a matching salt
+		if (salt_match_count != 0) {
```
Comment (lnkuiper) [inline src/execution/join_hashtable.cpp:579]: Can we move this part of the code that scans/compares to ensure chains have the same keys to a separate function with a descriptive name? Then this ~150 line function can be split into two functions, making the code more organized.

File: src/execution/join_hashtable.cpp [hunk 3]
```diff
@@ -243,70 +441,249 @@ idx_t JoinHashTable::PrepareKeys(DataChunk &keys, vector<TupleDataVectorFormat>
...
-		hash_map = buffer_manager.GetBufferAllocator().Allocate(capacity * sizeof(data_ptr_t));
+		hash_map = buffer_manager.GetBufferAllocator().Allocate(capacity * sizeof(aggr_ht_entry_t));
+		entries = reinterpret_cast<aggr_ht_entry_t *>(hash_map.get());
 	}
-	D_ASSERT(hash_map.GetSize() == capacity * sizeof(data_ptr_t));
+	D_ASSERT(hash_map.GetSize() == capacity * sizeof(aggr_ht_entry_t));
 
 	// initialize HT with all-zero entries
-	std::fill_n(reinterpret_cast<data_ptr_t *>(hash_map.get()), capacity, nullptr);
+	std::fill_n(entries, capacity, aggr_ht_entry_t(0));
```
Comment (lnkuiper) [inline src/execution/join_hashtable.cpp:680]: Nitpick: `aggr_ht_entry_t::GetEmptyEntry()`

File: src/include/duckdb/execution/aggregate_hashtable.hpp [hunk 4]
```diff
@@ -31,17 +31,34 @@ struct FlushMoveState;
 
 struct aggr_ht_entry_t { // NOLINT
```
Comment (lnkuiper) [inline src/include/duckdb/execution/aggregate_hashtable.hpp:32]: Since this class is now used by both the aggregate and join hashtables, could you move it to a separate file and rename it to `ht_entry_t`?

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Hi @gropaul, thanks for the PR! Great that you got this to work. I'm impressed that it even improved performance on TPC-H. There is a lot of added complexity to the code here, but I understand that this is necessary to get the "linear probing + equal key chaining" to work. I've left a few comments below requesting changes, but most are just to improve the code quality.

I'm interested to see how this affects many-to-many join performance, as this is where the new collision resolution should shine. Could you add a micro benchmark that performs a many-to-many join?

Something like this:
```sql
-- build side with 32k unique values, each repeating 4x
create table t1 as select 'verylargestring' || range % 32768 i from range(131072);
-- probe side with the same 32k unique values, each repeating 4096x
create table t2 as select 'verylargestring' || range % 32768 i from range(134217728);
-- join tables, should result in a count of 134217728 * 4 = 536870912
select count(*) from t1 join t2 using (i);
```
Every probed tuple will have a hit with a chain of 4. Your new implementation only has to perform 1 round of comparisons (string comparisons! which are expensive), instead of 4 (once for each tuple in the chain, like before). Your new implementation should be much faster.

I'd like to see the performance of this benchmark with and without your changes.

Finally, I'd like to know if you've tried tweaking the `PointerTableCapacity` function. Does the new collision resolution allow using a smaller pointer table, potentially reducing cache misses? In the aggregate hashtable we use a load factor of 1.5x, which makes it possible for up to 2/3rd of the entries to be filled. For the join hashtable, we use a load factor of 2x (hardcoded, not a constexpr like in the aggr ht), which allows only up to 1/2 of the entries to be filled, but if it improves performance, we might want to use 1.5x here as well.

**Comment by gropaul** [comment]
Hi @lnkuiper, thank you for the feedback, I will start implementing it today! 

For the benchmark, I wanted to ask you if you only want the results here in the PR, or should I also add them to the `benchmark/micro/join` in the repository or create a new benchmark directory there? 

In regards to the `PointerTableCapacity`, I played around a bit but the results didn't look very promising, although I didn't try it for scale factors of 30 and 100, where the cache misses will be most influential.I will run the TCP-H with different factors for the HT load factor and send you the results.

**Comment by lnkuiper** [comment]
> For the benchmark, I wanted to ask you if you only want the results here in the PR, or should I also add them to the benchmark/micro/join in the repository or create a new benchmark directory there?

@gropaul under `benchmark/micro/join` would be preferred

**Comment by gropaul** [comment]
Hi @lnkuiper, thanks for the tip with the microbenchmark, this case really suites the new implementation: I ran it on my mac book Pro (same as for the benchmarks above) using one thread and got the following results:

New Implementation: 3.733s, 
Current Implementation: 6.976s
Relative Speed Increase: 86.86%

##### Current Implementation
``` 
name	run	timing
benchmark/micro/join/hashjoin_dups_rhs.benchmark	1	6.994504
benchmark/micro/join/hashjoin_dups_rhs.benchmark	2	6.962721
benchmark/micro/join/hashjoin_dups_rhs.benchmark	3	6.985817
benchmark/micro/join/hashjoin_dups_rhs.benchmark	4	6.979697
benchmark/micro/join/hashjoin_dups_rhs.benchmark	5	6.957238
```

##### New Implementation
``` 
name	run	timing
benchmark/micro/join/hashjoin_dups_rhs.benchmark	1	3.724266
benchmark/micro/join/hashjoin_dups_rhs.benchmark	2	3.648539
benchmark/micro/join/hashjoin_dups_rhs.benchmark	3	3.791596
benchmark/micro/join/hashjoin_dups_rhs.benchmark	4	3.709812
benchmark/micro/join/hashjoin_dups_rhs.benchmark	5	3.793427
```

**Comment by lnkuiper** [comment]
Thanks for running the duplicates benchmark! The new code really shines there. I'm happy to see that.

If the load factor of 2 performs better than the load factor of 1.5, let's not change it. It was worth investigating.

I will review your changes shortly.

**Review by lnkuiper — APPROVED** [review]
Thanks for the changes! I've looked at it again, and it looks good now.

I've also ran TPC-H SF100 at a memory limit of 1GB and all was well so I'm confident that this has not affected the out-of-core join.

Note that it may take some time for this PR to be merged as the `feature` branch is currently low priority, as we are still doing bugfix releases for the `main` branch

**Comment by gropaul** [comment]
Thanks for the feedback and running the sf100 benchmark, if there is anything else to do let me know!

**Comment by lnkuiper** [comment]
Hi @gropaul, I just realized that we do not have to create scan the HT to create the de-duplicated chain in all cases.

If the join has `JoinType::SEMI` or `JoinType::ANTI`, we don't care about duplicates on the build side at all. Would it be possible to disable building the de-duplicated chain in that case?

File: src/execution/join_hashtable.cpp [hunk 5]
```diff
@@ -243,70 +445,267 @@ idx_t JoinHashTable::PrepareKeys(DataChunk &keys, vector<TupleDataVectorFormat>
...
+				data_ptr_t row_ptr_to_insert = lhs_row_locations[row_index];
+				data_ptr_t potential_collided_ptr =
+				    InsertRowToEntry<PARALLEL, true>(atomic_entry, row_ptr_to_insert, salt, ht->pointer_offset);
+
+				if (PARALLEL) {
+					// if the insertion was not successful, the entry was occupied in the meantime, so we have to
+					// compare the keys and insert the row to the next entry
+					if (potential_collided_ptr) {
+						// if the entry was occupied, we need to compare the keys and insert the row to the next entry
+						// we need to compare the keys and insert the row to the next entry
```
Comment (xuke-hat) [inline src/execution/join_hashtable.cpp:629]: Would it be better to check salt first and compare key only when salt is equal? By the way this line is duplicated.

Comment (gropaul) [reply src/execution/join_hashtable.cpp:629]: Hi @xuke-hat, thank you for the review! Good idea, I'll try to add the rows that fall in the case of line 628 to the remaining selection to process that in the next run of the outer loop, that way they will first compare the salt and then potentially the keys. 
Also, I'll address the code duplicate.

**Comment by lnkuiper** [comment]
Hi @gropaul,

Internally, we found a regression for our `RIGHT_SEMI` and `RIGHT_ANTI` join types, which only happens if there are MANY duplicates in the join hashtable. However, your PR fixes most of the problem! Thanks again :)

I've written a small patch that avoids a lot of duplicate work when there are many duplicates:
[right_semi_or_anti.patch](https://github.com/duckdb/duckdb/files/15245697/right_semi_or_anti.patch)

If you like, you could do `git apply right_semi_or_anti.patch` to apply my changes. Otherwise, I will add it later

**Comment by gropaul** [comment]
HI @lnkuiper, I merged the newest duckdb feature branch and applied your patch :) 

However, since merging the feature branch, I sometimes get segmentation faults when running large queries with multiple threads, like the `benchmark/micro/join/hashjoin_dups_rhs.benchmark`. I already was able to do some analysis, it occurred since the merge commit, so it has nothing to do with your patch. 

E.g. this triggers a seg fault in 90% of the tries:
```
/Users/paul/workspace/duckdb/build/release/benchmark/benchmark_runner benchmark/micro/join/hashjoin_dups_rhs.benchmark --threads=4 
```

I won't have the time till Monday to look into this, but I'll let you know as soon as I know more!

**Comment by lnkuiper** [comment]
Thanks for keeping this branch up-to-date @gropaul !

**Review by lnkuiper — APPROVED** [review]

**Comment by Mytherin** [comment]
Thanks! Looks great
